2022Swerve/src/main/java/frc/robot/subsystems/AutonSubsytem.java

```diff
@@ -0,0 +1,143 @@
+        FORWARD180; // Go forward 2 meters and rotate 180 degrees
+    }
+
+    private ShuffleboardTab configTab = Shuffleboard.getTab("Config");
+    private NetworkTableEntry delayEntry = configTab.add("Auton Delay", 0.0).getEntry();
+    private SendableChooser<AutonModes> modeChooser = new SendableChooser<>();
+
+    private DataLog datalog = DataLogManager.getLog();
+    private StringLogEntry trajectoryLog = new StringLogEntry(datalog, "/auton/trajectory"); //Logs x translation state output
+    private StringLogEntry commandLog = new StringLogEntry(datalog, "/auton/command"); //Logs x translation state output
+
+    private SwerveSubsystem swerveSubsystem;
+
+    private PIDController xController;
+    private PIDController yController;
+    private ProfiledPIDController rotationController;
+
+    private AutonModes autonMode;
+
+    public AutonSubsytem(SwerveSubsystem swerveSubsystem){
+        this.swerveSubsystem = swerveSubsystem;
+
+        //Add auton modes to chooser
+        for(AutonModes mode : AutonModes.values()){
+            modeChooser.addOption(mode.toString(), mode);
+        }
+        configTab.add("Auton mode", modeChooser);
+
+        //Define PID controllers for tracking trajectory
+        xController = new PIDController(AutonConstants.kXTranslationP, 0, 0);
+        yController = new PIDController(AutonConstants.kYTranslationP, 0, 0);
+        rotationController = new ProfiledPIDController(
+            AutonConstants.kRotationP, 0, 0, AutonConstants.kRotationalConstraints);
+        rotationController.enableContinuousInput(-Math.PI, Math.PI);
+    }
+
+    private Trajectory getTrajectory(String name){
+        Trajectory trajectory;
+        try {
+            Path trajectoryPath = Filesystem.getDeployDirectory().toPath().resolve("pathplanner/generatedJSON/" + name + ".wplib.json");
+            trajectory = TrajectoryUtil.fromPathweaverJson(trajectoryPath);
+        } catch (IOException ex) {
+            DriverStation.reportError("Unable to open trajectory: " + name, ex.getStackTrace());
+            trajectory = null;
+        }
+        return trajectory;
+    }
+
+    private SwerveControllerCommand followTrajectory(String name){
+        trajectoryLog.append(name);
+        return new SwerveControllerCommand(
+            getTrajectory(name),
+            swerveSubsystem::getPose, 
+            DriveConstants.kDriveKinematics, 
+            xController, 
+            yController, 
+            rotationController, 
+            swerveSubsystem::setModuleStates, 
+            swerveSubsystem
+        );
+    }
+
+    private Command resetOdometry(String initialTrajectory){
+        //Resets odometry to the initial position of the given trajectory
+        return new InstantCommand(() -> swerveSubsystem.resetOdometry(getTrajectory(initialTrajectory).getInitialPose()));
+    }
+
+    private Command delay(double seconds){
+        commandLog.append("Wait " + seconds);
```

Comment:
Been thinking about this.  When does the command log get run?  
When does the waitcommand happen?

Not necessarily at the same time.  Maybe use .initialize?

Reply:
I don't understand

Is the problem that the logs aren't correct to the nanosecond? Or is there something I'm missing

Reply:
I think he means the log happens before the WaitCommand this method returns actually gets scheduled. i.e. if someone called ```delay(10)```, but then never used that command it would still log "Wait 10" even though no waiting happened.

Two ideas to fix:
1. Add ``` .beforeStarting(() -> commandLog.append("Wait " + seconds))``` to the returned command instead.
2. Instead of a method, create a child class WaitCommandWithLog and just override the initialize() method to do
```
super();
commandLog.append("Wait " + seconds);
```


P.S. might want to add the unit label to the string ```"Wait " + seconds + " seconds." ```